File: valluvan/DatabaseManager.swift
import Foundation
import SQLite // Make sure this import is correct
import UIKit // Add this import for NSAttributedString

public struct DatabaseSearchResult: Identifiable {
    public let id: Int
    public let heading: String
    public let subheading: String
    public let content: String
    public let explanation: String
    public let kuralId: Int
    
    public init(heading: String, subheading: String, content: String, explanation: String, kuralId: Int) {
        self.id = kuralId // Use kuralId as the unique identifier
        self.heading = heading
        self.subheading = subheading
        self.content = content
        self.explanation = explanation
        self.kuralId = kuralId
    }
}

public class DatabaseManager {
    public static let shared = DatabaseManager()
    private var db: Connection?
    
    private init() {
        do {
            if let path = Bundle.main.path(forResource: "data", ofType: "sqlite") {
                db = try Connection(path)
            } else {
                print("Database file not found in the main bundle.")
                print("Searched for 'data.sqlite' in: \(Bundle.main.bundlePath)")
                
                // List contents of the bundle for debugging
                let fileManager = FileManager.default
                if let enumerator = fileManager.enumerator(atPath: Bundle.main.bundlePath) {
                    print("Contents of the main bundle:")
                    while let filePath = enumerator.nextObject() as? String {
                        print(filePath)
                    }
                }
            }
        } catch {
            print("Unable to connect to database: \(error)")
        }
    }

    
    public func getIyals(for pal: String, language: String) -> [String] {
        let tirukkuralTable = Table("tirukkural")
        let kuralId = Expression<Int>("kno")
        let palExpr = language == "Tamil" ? Expression<String>("pal") : Expression<String>("title")
        let iyalExpr = language == "Tamil" ? Expression<String>("iyal") : Expression<String>("heading")
        
        var iyals: [String] = []
        
        do {
            let query = tirukkuralTable
                .select(iyalExpr)
                .filter(palExpr == pal)
                .group(iyalExpr)
                .order(kuralId)
             
            for row in try db!.prepare(query) {
                iyals.append(row[iyalExpr])
            }
        } catch {
            print("Error fetching iyals: \(error)")
        } 
            
        return iyals
    }
 
    public func getAdhigarams(for iyal: String, language: String) -> ([String], [Int], [String], [String]) {
        let tirukkuralTable = Table("tirukkural")
        let kuralId = Expression<Int>("kno")
        let iyalExpr = language == "Tamil" ? Expression<String>("iyal") : Expression<String>("heading")
        let adhigaramExpr = language == "Tamil" ? Expression<String>("tchapter") : Expression<String>("chapter")
        let adhigaramSongExpr = Expression<String>("tchapter") 
        var adhigarams: [String] = []
        var kuralIds: [Int] = []
        var adhigaramSongs: [String] = []
        var originalAdhigarams: [String] = []
        do {
            let query = tirukkuralTable
                .select(adhigaramExpr, kuralId, adhigaramSongExpr)
                .filter(iyalExpr == iyal)
                .group(adhigaramExpr)
                .order(kuralId)
            
            for row in try db!.prepare(query) {
                adhigarams.append(row[adhigaramExpr])   
                originalAdhigarams.append(row[adhigaramExpr])
                kuralIds.append(row[kuralId])
                adhigaramSongs.append(row[adhigaramSongExpr])
            }
        } catch {
            print("Error fetching adhigarams: \(error)")
        }

        return (adhigarams, kuralIds, adhigaramSongs, originalAdhigarams)
    } 

    public func getSingleLine(for adhigaram: String, language: String) -> [String] {
        let tirukkuralTable = Table("tirukkural")
        let kuralId = Expression<Int>("kno") 
        let adhigaramExpr = Expression<String>("chapter")
        
        let firstLineExpr = Expression<String>(language) 
         
        var kurals: [String] = []
        do {
            let query = tirukkuralTable
                .select(kuralId, firstLineExpr)   
                .filter(adhigaramExpr == adhigaram)
                .order(kuralId)  

            for row in try db!.prepare(query) {
                kurals.append(String(row[kuralId]) + " " + row[firstLineExpr]) 
            }
        } catch {
            print("Error fetching first line: \(error)")
        }
        return kurals
    }

    public func getFirstLine(for adhigaram: String, language: String) -> [String] {
        let tirukkuralTable = Table("tirukkural")
        let kuralId = Expression<Int>("kno") 
        let adhigaramExpr = language == "Tamil" ? Expression<String>("tchapter") : Expression<String>("chapter")
        
        let firstLineExpr: Expression<String>
        let secondLineExpr: Expression<String>
        
        switch language {
        case "Tamil":
            firstLineExpr = Expression<String>("firstline")
            secondLineExpr = Expression<String>("secondline")
        case "telugu":
            firstLineExpr = Expression<String>("telugu1")
            secondLineExpr = Expression<String>("telugu2")
        case "hindi":
            firstLineExpr = Expression<String>("hindi1")
            secondLineExpr = Expression<String>("hindi2")
        default:
            firstLineExpr = Expression<String>("efirstline")
            secondLineExpr = Expression<String>("esecondline")
        }
        
        var kurals: [String] = []
        do {
            let query = tirukkuralTable
                .select(kuralId, firstLineExpr, secondLineExpr)   
                .filter(adhigaramExpr == adhigaram)
                .order(kuralId)  

            for row in try db!.prepare(query) {
                kurals.append(String(row[kuralId]) + " " + row[firstLineExpr])
                kurals.append(row[secondLineExpr])
            }
        } catch {
            print("Error fetching first line: \(error)")
        }
        return kurals
    }
    
    func getExplanation(for kuralId: Int, language: String) -> NSAttributedString {
        let tirukkuralTable = Table("tirukkural")
        let kuralIdExpr = Expression<Int>("kno")
        let explanationExpr: Expression<String>
        let manaExplanationExpr: Expression<String>
        let pariExplanationExpr: Expression<String>
        let varaExplanationExpr: Expression<String>
        let popsExplanationExpr: Expression<String>
        let muniExplanationExpr: Expression<String>
        var query: Table
        var attributedExplanation = NSMutableAttributedString()

        switch language {
        case "Tamil":
            explanationExpr = Expression<String>("kalaignar")
            manaExplanationExpr = Expression<String>("manakudavar")
            pariExplanationExpr = Expression<String>("parimelazhagar")
            varaExplanationExpr = Expression<String>("varadarajanar")
            popsExplanationExpr = Expression<String>("salomon")
            muniExplanationExpr = Expression<String>("munisamy")
            query = tirukkuralTable
                .select(explanationExpr, manaExplanationExpr, pariExplanationExpr, varaExplanationExpr, popsExplanationExpr, muniExplanationExpr)
                .filter(kuralIdExpr == kuralId) 
            do {                
                if let row = try db!.pluck(query) {
                    let boldAttributes: [NSAttributedString.Key: Any] = [.font: UIFont.boldSystemFont(ofSize: UIFont.systemFontSize)]
                    
                    appendExplanation(to: &attributedExplanation, title: "கலைஞர் விளக்கம்: ", content: row[explanationExpr], boldAttributes: boldAttributes)
                    appendExplanation(to: &attributedExplanation, title: "மணக்குடவர் விளக்கம்: ", content: row[manaExplanationExpr], boldAttributes: boldAttributes)
                    appendExplanation(to: &attributedExplanation, title: "பரிமேலழகர் விளக்கம்: ", content: row[pariExplanationExpr], boldAttributes: boldAttributes)
                    appendExplanation(to: &attributedExplanation, title: "மு. வரதராசன் விளக்கம்: ", content: row[varaExplanationExpr], boldAttributes: boldAttributes)
                    appendExplanation(to: &attributedExplanation, title: "சாலமன் பாப்பையா விளக்கம்: ", content: row[popsExplanationExpr], boldAttributes: boldAttributes)    
                    appendExplanation(to: &attributedExplanation, title: "வீ. முனிசாமி விளக்கம்: ", content: row[muniExplanationExpr], boldAttributes: boldAttributes, isLast: true)
                }
            } catch {
                print("Error fetching Tamil explanation: \(error)")
            }   
        default:
            explanationExpr = Expression<String>("explanation") 
            do {                 
                query = tirukkuralTable
                    .select(explanationExpr)
                    .filter(kuralIdExpr == kuralId) 
                
                if let row = try db!.pluck(query) {
                    attributedExplanation = NSMutableAttributedString(string: row[explanationExpr])
                }
            } catch {
                print("Error fetching explanation: \(error)")
            } 
        }
        return attributedExplanation
    }

    private func appendExplanation(to attributedString: inout NSMutableAttributedString, title: String, content: String, boldAttributes: [NSAttributedString.Key: Any], isLast: Bool = false) {
        attributedString.append(NSAttributedString(string: title, attributes: boldAttributes))
        attributedString.append(NSAttributedString(string: content))
        if !isLast {
            attributedString.append(NSAttributedString(string: "\n\n"))
        }
    }

    func searchContent(query: String, language: String) -> [DatabaseSearchResult] {
        var results: [DatabaseSearchResult] = []
        let searchQuery: String
        let searchPattern = "%\(query)%"

        if language != "English" && language != "telugu" && language != "hindi" {
            searchQuery = """
                SELECT "kno", "heading", "chapter", "efirstline", "esecondline", "explanation", "\(language)"
                FROM tirukkural
                WHERE "heading" LIKE ? OR "chapter" LIKE ? OR "efirstline" LIKE ? OR "esecondline" LIKE ? OR "explanation" LIKE ? OR "\(language)" LIKE ?
                LIMIT 20
            """
            
            do {
                let rows = try db!.prepare(searchQuery, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern)
                for row in rows {
                    let result = DatabaseSearchResult(
                        heading: row[1] as? String ?? "",
                        subheading: row[2] as? String ?? "",
                        content: "\(row[6] as? String ?? "")\n\(row[3] as? String ?? "")\n\(row[4] as? String ?? "")",
                        explanation: row[5] as? String ?? "",
                        kuralId: Int(row[0] as? Int64 ?? 0)
                    )
                    results.append(result)
                }
            } catch {
                print("Error searching content: \(error.localizedDescription)")
            }    
        } else {
            if language == "English" {
                searchQuery = """
                    SELECT "kno", "heading", "chapter", "efirstline", "esecondline", "explanation"
                    FROM tirukkural
                    WHERE "heading" LIKE ? OR "chapter" LIKE ? OR "efirstline" LIKE ? OR "esecondline" LIKE ? OR "explanation" LIKE ?
                    LIMIT 20
                """
            } else {
                searchQuery = """
                    SELECT "kno", "heading", "chapter", "\(language)1", "\(language)2", "explanation"
                    FROM tirukkural
                    WHERE "heading" LIKE ? OR "chapter" LIKE ? OR "\(language)1" LIKE ? OR "\(language)2" LIKE ? OR "explanation" LIKE ?
                    LIMIT 20
                """
            }
            do {
                let rows = try db!.prepare(searchQuery, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern)
                for row in rows {
                    let result = DatabaseSearchResult(
                        heading: row[1] as? String ?? "",
                        subheading: row[2] as? String ?? "",
                        content: "\(row[3] as? String ?? "")\n\(row[4] as? String ?? "")",
                        explanation: row[5] as? String ?? "",
                        kuralId: Int(row[0] as? Int64 ?? 0)
                    )
                    results.append(result)
                }
            } catch {
                print("Error searching content: \(error.localizedDescription)")
            }
        }

        return results
    }


    func searchTamilContent(query: String) -> [DatabaseSearchResult] {
        var results: [DatabaseSearchResult] = []
        let searchQuery = """
            SELECT "kno", "iyal", "tchapter", "firstline", "secondline", "manakudavar", "parimelazhagar", "varadarajanar", "kalaignar", "salomon", "munisamy", "efirstline", "esecondline", "explanation"
            FROM tirukkural
            WHERE "iyal" LIKE ? OR "tchapter" LIKE ? OR "firstline" LIKE ? OR "secondline" LIKE ? OR "manakudavar" LIKE ? OR "parimelazhagar" LIKE ? OR "varadarajanar" LIKE ? OR "kalaignar" LIKE ? OR "salomon" LIKE ? OR "munisamy" LIKE ? OR "efirstline" LIKE ? OR "esecondline" LIKE ? OR "explanation" LIKE ?
            LIMIT 20
        """
        let searchPattern = "%\(query)%"

        do {
            let rows = try db!.prepare(searchQuery, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern, searchPattern)
            for row in rows {
                let result = DatabaseSearchResult(
                    heading: row[1] as? String ?? "",
                    subheading: row[2] as? String ?? "",
                    content: "\(row[3] as? String ?? "")\n\(row[4] as? String ?? "")",
                    explanation: row[8] as? String ?? "",
                    kuralId: Int(row[0] as? Int64 ?? 0)
                )
                results.append(result)
            }
        } catch {
            print("Error searching Tamil content: \(error.localizedDescription)")
        }

        return results
    }

    func getKuralById(_ kuralId: Int, language: String) -> DatabaseSearchResult? {
        let tirukkuralTable = Table("tirukkural")
        let kuralIdExpr = Expression<Int>("kno")
        let headingExpr = Expression<String>("heading")
        let subheadingExpr = Expression<String>("chapter")
        let contentExpr1 = Expression<String>("firstline")
        let contentExpr2 = Expression<String>("secondline")
        let tfirstLineExpr = Expression<String>("telugu1")
        let tsecondLineExpr = Expression<String>("telugu2")
        let hfirstLineExpr = Expression<String>("hindi1")
        let hsecondLineExpr = Expression<String>("hindi2")
        let explanationExpr: Expression<String>
        
        switch language {
        case "Tamil":
            explanationExpr = Expression<String>("kalaignar")
        case "English", "hindi", "telugu":
            explanationExpr = Expression<String>("explanation")
        default:
            explanationExpr = Expression<String>(language)
        }

        do {
            let query = tirukkuralTable
                .select(headingExpr, subheadingExpr, contentExpr1, contentExpr2, tfirstLineExpr, tsecondLineExpr, hfirstLineExpr, hsecondLineExpr, explanationExpr)
                .filter(kuralIdExpr == kuralId)

            if let row = try db!.pluck(query) {
                return DatabaseSearchResult(
                    heading: row[headingExpr],
                    subheading: row[subheadingExpr],
                    content: language == "telugu" ? "\(row[tfirstLineExpr])\n\(row[tsecondLineExpr])" : language == "hindi" ? "\(row[hfirstLineExpr])\n\(row[hsecondLineExpr])" : "\(row[contentExpr1])\n\(row[contentExpr2])",
                    explanation: row[explanationExpr],
                    kuralId: kuralId
                )
            }
        } catch {
            print("Error fetching Kural by ID: \(error)")
        }

        return nil
    }
}

File: app/src/main/java/com/example/valluvan/DatabaseManager.kt
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.text.SpannableString
import android.text.style.StyleSpan
import android.graphics.Typeface

class DatabaseManager private constructor(context: Context) {
    private var db: SQLiteDatabase? = null

    init {
        try {
            val dbPath = context.getDatabasePath("data.sqlite").absolutePath
            db = SQLiteDatabase.openDatabase(dbPath, null, SQLiteDatabase.OPEN_READONLY)
            println("Connected to database at path: $dbPath")
        } catch (e: Exception) {
            println("Unable to connect to database: ${e.message}")
        }
    }

    fun getIyals(pal: String, language: String): List<String> {
        val iyals = mutableListOf<String>()
        val palColumn = if (language == "Tamil") "பால்" else "English Title"
        val iyalColumn = if (language == "Tamil") "இயல்" else "English Heading"
        
        val query = """
            SELECT DISTINCT $iyalColumn
            FROM tirukkural
            WHERE $palColumn = ?
            ORDER BY திருக்குறள்
        """.trimIndent()

        db?.rawQuery(query, arrayOf(pal))?.use { cursor ->
            while (cursor.moveToNext()) {
                iyals.add(cursor.getString(0))
            }
        }

        return iyals
    }

    fun getAdhigarams(iyal: String, language: String): List<String> {
        val adhigarams = mutableListOf<String>()
        val iyalColumn = if (language == "Tamil") "இயல்" else "English Heading"
        val adhigaramColumn = if (language == "Tamil") "அதிகாரம்" else "English Chapter"
        
        val query = """
            SELECT DISTINCT $adhigaramColumn
            FROM tirukkural
            WHERE $iyalColumn = ?
            ORDER BY திருக்குறள்
        """.trimIndent()

        db?.rawQuery(query, arrayOf(iyal))?.use { cursor ->
            while (cursor.moveToNext()) {
                adhigarams.add(cursor.getString(0))
            }
        }

        return adhigarams
    }

    fun getSingleLine(adhigaram: String, language: String): List<String> {
        val kurals = mutableListOf<String>()
        val query = """
            SELECT திருக்குறள், $language
            FROM tirukkural
            WHERE `English Chapter` = ?
            ORDER BY திருக்குறள்
        """.trimIndent()

        db?.rawQuery(query, arrayOf(adhigaram))?.use { cursor ->
            while (cursor.moveToNext()) {
                val kuralId = cursor.getInt(0)
                val line = cursor.getString(1)
                kurals.add("$kuralId $line")
            }
        }

        return kurals
    }

    fun getFirstLine(adhigaram: String, language: String): List<List<String>> {
        val kurals = mutableListOf<List<String>>()
        val (firstLineColumn, secondLineColumn) = when (language) {
            "Tamil" -> Pair("First Line", "Second Line")
            "Telugu" -> Pair("Telugu 1", "Telugu 2")
            "Hindi" -> Pair("Hindi 1", "Hindi 2")
            else -> Pair("First Line English", "Second Line English")
        }

        val query = """
            SELECT திருக்குறள், `$firstLineColumn`, `$secondLineColumn`
            FROM tirukkural
            WHERE `English Chapter` = ?
            ORDER BY திருக்குறள்
        """.trimIndent()

        db?.rawQuery(query, arrayOf(adhigaram))?.use { cursor ->
            while (cursor.moveToNext()) {
                val kuralId = cursor.getInt(0)
                val firstLine = "${cursor.getString(1)}"
                val secondLine = cursor.getString(2)
                kurals.add(listOf("$kuralId $firstLine", secondLine))
            }
        }

        return kurals
    }

    fun getExplanation(kuralId: Int, language: String): SpannableString {
        val explanationBuilder = StringBuilder()
        val query = if (language == "Tamil") {
            """
            SELECT கலைஞர், மணக்குடவர், பரிமேலழகர், `மு. வரதராசன்`, `சாலமன் பாப்பையா`, `வீ. முனிசாமி`
            FROM tirukkural
            WHERE திருக்குறள் = ?
            """.trimIndent()
        } else {
            "SELECT Explanation FROM tirukkural WHERE திருக்குறள் = ?"
        }

        db?.rawQuery(query, arrayOf(kuralId.toString()))?.use { cursor ->
            if (cursor.moveToFirst()) {
                if (language == "Tamil") {
                    val titles = listOf("கலைஞர்", "மணக்குடவர்", "பரிமேலழகர்", "மு. வரதராசன்", "சாலமன் பாப்பையா", "வீ. முனிசாமி")
                    for (i in 0 until cursor.columnCount) {
                        explanationBuilder.append("${titles[i]}: ${cursor.getString(i)}\n\n")
                    }
                } else {
                    explanationBuilder.append(cursor.getString(0))
                }
            }
        }

        val spannableString = SpannableString(explanationBuilder.toString())
        if (language == "Tamil") {
            var startIndex = 0
            while (true) {
                val colonIndex = explanationBuilder.indexOf(": ", startIndex)
                if (colonIndex == -1) break
                spannableString.setSpan(StyleSpan(Typeface.BOLD), startIndex, colonIndex, 0)
                startIndex = explanationBuilder.indexOf("\n\n", colonIndex) + 2
            }
        }

        return spannableString
    }

    fun searchContent(query: String): List<SearchResult> {
        val results = mutableListOf<SearchResult>()
        val searchQuery = """
            SELECT திருக்குறள், `English Heading`, `English Chapter`, `First Line English`, `Second Line English`, Explanation
            FROM tirukkural
            WHERE `English Heading` LIKE ? OR `English Chapter` LIKE ? OR `First Line English` LIKE ? OR `Second Line English` LIKE ? OR Explanation LIKE ?
            LIMIT 20
        """.trimIndent()
        val searchPattern = "%$query%"

        db?.rawQuery(searchQuery, Array(5) { searchPattern })?.use { cursor ->
            while (cursor.moveToNext()) {
                val result = SearchResult(
                    kuralId = cursor.getInt(0),
                    heading = cursor.getString(1),
                    subheading = cursor.getString(2),
                    content = "${cursor.getString(3)}\n${cursor.getString(4)}",
                    explanation = cursor.getString(5)
                )
                results.add(result)
            }
        }

        return results
    }

    companion object {
        @Volatile
        private var instance: DatabaseManager? = null

        fun getInstance(context: Context): DatabaseManager {
            return instance ?: synchronized(this) {
                instance ?: DatabaseManager(context).also { instance = it }
            }
        }
    }
}

data class SearchResult(
    val kuralId: Int,
    val heading: String,
    val subheading: String,
    val content: String,
    val explanation: String
)